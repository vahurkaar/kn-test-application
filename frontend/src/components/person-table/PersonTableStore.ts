import { action, makeObservable, observable } from 'mobx';
import Person from '../../models/Person';
import Page from '../../models/Page';
import { debounce } from 'debounce';
import PersonApi, { PersonsResponse } from '../../api/PersonApi';
import { Debounce } from '../../../typings';

export default class PersonTableStore {
  @observable people: Person[] = [];
  @observable page: Page;
  @observable nameFilter: string = '';
  @observable isInitialized: boolean = false;
  @observable hasErrors: boolean = false;

  constructor() {
    makeObservable(this);
  }

  @action
  init = async () => {
    await this.findPeople();
    this.setInitialized();
  }

  @action
  findPeople = async (page?: Page) => {
    return PersonApi.findPersonsByName(this.nameFilter, page).then(this.setState).catch(this.setErrors);
  }

  @action
  onSelectPage = async (page: Page) => {
    await this.findPeople(page);
  }

  @action
  setState = ({ people, page }: PersonsResponse) => {
    this.people = people;
    this.page = page;
  }

  @action
  onChangePersonFilter = async (name?: string) => {
    this.nameFilter = name?.trim();
    await this.nameFilterChangeCallback();
  }

  @action
  onPersonFilterBlur = () => {
    this.nameFilterChangeCallback.flush();
  }

  @action
  setInitialized = () => {
    this.isInitialized = true;
  }

  @action
  setErrors = () => {
    this.hasErrors = true;
  }

  nameFilterChangeCallback: Debounce<(page?: Page) => void> = debounce(this.findPeople, 2000);
}
