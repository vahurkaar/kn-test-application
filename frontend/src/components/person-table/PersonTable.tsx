import { Alert, Spinner, Table } from 'react-bootstrap';
import React, { useEffect } from 'react';
import { observer, useLocalObservable } from 'mobx-react';
import PersonTableStore from './PersonTableStore';
import TablePagination from '../shared/TablePagination';
import Person from '../../models/Person';
import SearchField from '../shared/SearchField';

const PersonTable = observer(() => {
  const store = useLocalObservable(() => new PersonTableStore());
  const { people, page, hasErrors, isInitialized, onSelectPage, onChangePersonFilter, onPersonFilterBlur } = store;

  const getPersonIndex = (index: number) => page.size * page.number + index + 1 ;
  const getPersonName = (person: Person) => person.name + (person.suffix ? ` (${person.suffix})` : '');

  useEffect(() => {
    store.init();
  }, [store]);

  if (hasErrors) {
    return <Alert variant="danger">System error!</Alert>;
  }

  if (!isInitialized) {
    return (
      <Spinner animation="border" role="status">
        <span className="visually-hidden">Loading...</span>
      </Spinner>
    );
  }

  return (
    <>
      <h3>Persons</h3>

      <SearchField label="Search by name" onChange={onChangePersonFilter} onBlur={onPersonFilterBlur} />
      <TablePagination page={page} onSelect={onSelectPage} />
      <Table striped bordered hover>
        <thead>
        <tr>
          <th>#</th>
          <th>Name</th>
          <th>Image</th>
        </tr>
        </thead>
        <tbody>
        {people.map((person: Person, index: number) => (
          <tr key={`person-${index}`}>
            <td>{getPersonIndex(index)}</td>
            <td>{getPersonName(person)}</td>
            <td><img height={50} src={person.url} alt="" /></td>
          </tr>
        ))}
        </tbody>
      </Table>
      <TablePagination page={page} onSelect={onSelectPage} />
    </>
  )
});

export default PersonTable;