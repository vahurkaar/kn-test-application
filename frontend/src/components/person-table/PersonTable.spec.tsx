import PersonTable from './PersonTable';
import { render } from '@testing-library/react';
import React from 'react';
import PersonTableStore from './PersonTableStore';

jest.mock('./PersonTableStore', () => ({
  __esModule: true,
  default: jest.fn()
}));

jest.mock('../shared/TablePagination', () => 'table-pagination');
jest.mock('../shared/SearchField', () => 'search-field');

describe('PersonTable', () => {
  const create = (props: Partial<PersonTableStore> = {}) => {
    const store: Partial<PersonTableStore> = {
      init: jest.fn(),
      isInitialized: true,
      people: [{
        name: 'name',
        suffix: 'suffix',
        url: 'url'
      }, {
        name: 'name without suffix',
        url: 'url-2'
      }],
      page: {
        number: 0,
        totalPages: 10,
        totalElements: 100,
        size: 10
      },
      ...props
    };

    // @ts-ignore
    PersonTableStore.mockImplementation(() => {
      return {
        ...store
      };
    });

    const rendered = render(<PersonTable/>);
    return { rendered, store };
  };

  it('should render', () => {
    const { rendered: { container }, store } = create();
    expect(container).toMatchSnapshot();
    expect(store.init).toHaveBeenCalled();
  });

  it('should render empty list', () => {
    const { rendered: { container }, store } = create({ people: [] });
    expect(container).toMatchSnapshot();
    expect(store.init).toHaveBeenCalled();
  });

  it('should render spinner', () => {
    const { rendered: { container }, store } = create({ isInitialized: false });
    expect(container).toMatchSnapshot();
    expect(store.init).toHaveBeenCalled();
  });

  it('should render alert', () => {
    const { rendered: { container } } = create({ hasErrors: true });
    expect(container).toMatchSnapshot();
  });
});