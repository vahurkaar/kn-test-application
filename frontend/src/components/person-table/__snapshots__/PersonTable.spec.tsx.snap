// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonTable should render 1`] = `
<div>
  <h3>
    Persons
  </h3>
  <search-field
    label="Search by name"
  />
  <table-pagination
    page="[object Object]"
  />
  <table
    class="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          Name
        </th>
        <th>
          Image
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          1
        </td>
        <td>
          name (suffix)
        </td>
        <td>
          <img
            alt=""
            height="50"
            src="url"
          />
        </td>
      </tr>
      <tr>
        <td>
          2
        </td>
        <td>
          name without suffix
        </td>
        <td>
          <img
            alt=""
            height="50"
            src="url-2"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <table-pagination
    page="[object Object]"
  />
</div>
`;

exports[`PersonTable should render alert 1`] = `
<div>
  <div
    class="fade alert alert-danger show"
    role="alert"
  >
    System error!
  </div>
</div>
`;

exports[`PersonTable should render empty list 1`] = `
<div>
  <h3>
    Persons
  </h3>
  <search-field
    label="Search by name"
  />
  <table-pagination
    page="[object Object]"
  />
  <table
    class="table table-striped table-bordered table-hover"
  >
    <thead>
      <tr>
        <th>
          #
        </th>
        <th>
          Name
        </th>
        <th>
          Image
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
  <table-pagination
    page="[object Object]"
  />
</div>
`;

exports[`PersonTable should render spinner 1`] = `
<div>
  <div
    class="spinner-border"
    role="status"
  >
    <span
      class="visually-hidden"
    >
      Loading...
    </span>
  </div>
</div>
`;
